components/filecoin/rs/src/signature.rs

```diff
@@ -0,0 +1,187 @@
+
+    let cid_hashed = Params::new()
+        .hash_length(HASH_LENGTH)
+        .to_state()
+        .update(CID_PREFIX)
+        .update(message_hashed.as_bytes())
+        .finalize();
+
+    cid_hashed.as_bytes().try_into()
+}
+
+fn transaction_sign_secp56k1_raw(
+    unsigned_message_api: &UnsignedMessage,
+    private_key: &PrivateKey,
+) -> Result<SignatureSECP256K1, SignerError> {
+    let message_cbor = unsigned_message_api.marshal_cbor()?;
+
+    let secret_key = libsecp256k1::SecretKey::parse_slice(&private_key.0)?;
+
+    let cid_hashed = get_digest(message_cbor.as_ref())?;
+
+    let message_digest = Message::parse_slice(&cid_hashed)?;
+
+    let (signature_rs, recovery_id) = sign(&message_digest, &secret_key);
+
+    let mut signature = SignatureSECP256K1([0; SIGNATURE_RECOVERY_SIZE]);
+    signature.0[..64].copy_from_slice(&signature_rs.serialize()[..]);
+    signature.0[64] = recovery_id.serialize();
+
+    Ok(signature)
+}
+
+fn transaction_sign_bls_raw(
+    unsigned_message: &UnsignedMessage,
+    private_key: &PrivateKey,
+) -> Result<SignatureBLS, SignerError> {
+    let sk = bls_signatures::PrivateKey::from_bytes(&private_key.0)?;
+
+    //sign the message's signing bytes
+    let signed = sk.sign(unsigned_message.to_signing_bytes());
+    let v = signed.as_bytes();
+    if v.len() != BLS_SIGNATURE_SIZE {
+        return Err(SignerError::GenericString("Invalid Signature Length".to_string()));
+    }
+    let mut signature = SignatureBLS { 0: [0; BLS_SIGNATURE_SIZE] };
+    signature.0.copy_from_slice(&v[..BLS_SIGNATURE_SIZE]);
+
+    Ok(signature)
+}
+
+/// Sign a transaction and return a raw signature (RSV format).
+///
+/// # Arguments
+///
+/// * `unsigned_message_api` - an unsigned filecoin message
+/// * `private_key` - a `PrivateKey`
+///
+pub fn transaction_sign_raw(
+    unsigned_message_api: &UnsignedMessage,
+    private_key: &PrivateKey,
+) -> Result<Signature, SignerError> {
+    // the `from` address protocol let us know which signing scheme to use
+    match unsigned_message_api.from.protocol() {
+        Protocol::Secp256k1 => {
+            let signed = transaction_sign_secp56k1_raw(unsigned_message_api, private_key)?;
+            Ok(Signature::SignatureSECP256K1(signed))
+        }
+        Protocol::BLS => {
+            let signed = transaction_sign_bls_raw(unsigned_message_api, private_key)?;
+            Ok(Signature::SignatureBLS(signed))
+        }
+        _ => Err(SignerError::GenericString("Unknown signing protocol".to_string())),
+    }
+}
+
+pub fn transaction_sign(transaction: &str, private_key_base64: &str) -> String {
+    let mut de = serde_json::Deserializer::from_str(transaction);
+    let private_key = base64::decode(private_key_base64);
+    if private_key.is_err() {
+        return String::new();
+    }
+    let mut sk = PrivateKey { 0: [0; SECRET_KEY_SIZE] };
+    sk.0.copy_from_slice(&private_key.unwrap()[..SECRET_KEY_SIZE]);
```

Comment:
@rillian can you double check this `unwrap()` call here? I've always been told using unwrap in production is a bad idea because it can panic exit.

Reply:
Also is it ever possible for the stack trace dropped to return the private_key material here? I'd be concerned that if this does panic that a crash log would end up containing private_key material here

Reply:
@kdenhartog this unwrap is guarded by the `private_key.is_err()` check few lines up, which returns early so it can't be triggered in practice. (At least not without memory corruption.)

The code is brittle in that the guard and could be changed without updating the unwrap. It can be rewritten as a match or an else-clause for better robustness in subsequent edits, but that adds an indentation level, so I judged it acceptable since the flow is straightforward here. Especially since this is FFI glue where we can't use `?` for early returns.

If a panic could actually occur here, I think you're correct that it would leave the key material on the stack where it could serialized in a crash report.

Reply:
Ok, thanks for weighing in on this here. From a security PoV my preference is to refactor out just so we can't accidentally stack trace a private key. My thinking is that the tradeoff of updating this upfront is much easier than an accidental change/odd memory issue in the future which starts leaking keys and then we have to report security breaches and go through key rotation exercises etc. @spylogsster thoughts on refactoring out that unwrap?

Reply:
ok, please check this commit https://github.com/brave/brave-core/pull/13061/commits/4adc8d63ed0d15b44cce9e7678c4fb2cecea3314